chips/stm32f4xx/src/spi.rs

```diff
@@ -0,0 +1,500 @@
+        self.clock.enable();
+    }
+
+    pub fn disable_clock(&self) {
+        self.clock.disable();
+    }
+
+    pub fn set_dma(&self, tx_dma: TxDMA<'a>, rx_dma: RxDMA<'a>) {
+        self.tx_dma.set(tx_dma.0);
+        self.rx_dma.set(rx_dma.0);
+    }
+
+    pub fn handle_interrupt(&self) {
+        // Used only during debugging. Since we use DMA, we do not enable SPI
+        // interrupts during normal operations
+    }
+
+    // for use by dma1
+    pub fn get_address_dr(&self) -> u32 {
+        &self.registers.dr as *const ReadWrite<u32, DR::Register> as u32
+    }
+
+    fn set_active_slave(&self, slave_pin: PinId) {
+        self.active_slave.set(slave_pin);
+    }
+
+    fn set_cr<F>(&self, f: F)
+    where
+        F: FnOnce(),
+    {
+        self.registers.cr1.modify(CR1::SPE::CLEAR);
+        f();
+        self.registers.cr1.modify(CR1::SPE::SET);
+    }
+
+    // IdleLow  = CPOL = 0
+    // IdleHigh = CPOL = 1
+    fn set_polarity(&self, polarity: ClockPolarity) {
+        self.set_cr(|| match polarity {
+            ClockPolarity::IdleLow => self.registers.cr1.modify(CR1::CPOL::CLEAR),
+            ClockPolarity::IdleHigh => self.registers.cr1.modify(CR1::CPOL::SET),
+        });
+    }
+
+    fn get_polarity(&self) -> ClockPolarity {
+        if !self.registers.cr1.is_set(CR1::CPOL) {
+            ClockPolarity::IdleLow
+        } else {
+            ClockPolarity::IdleHigh
+        }
+    }
+
+    // SampleLeading  = CPHA = 0
+    // SampleTrailing = CPHA = 1
+    fn set_phase(&self, phase: ClockPhase) {
+        self.set_cr(|| match phase {
+            ClockPhase::SampleLeading => self.registers.cr1.modify(CR1::CPHA::CLEAR),
+            ClockPhase::SampleTrailing => self.registers.cr1.modify(CR1::CPHA::SET),
+        });
+    }
+
+    fn get_phase(&self) -> ClockPhase {
+        if !self.registers.cr1.is_set(CR1::CPHA) {
+            ClockPhase::SampleLeading
+        } else {
+            ClockPhase::SampleTrailing
+        }
+    }
+
+    fn enable_tx(&self) {
+        self.registers.cr2.modify(CR2::TXDMAEN::SET);
+    }
+
+    fn disable_tx(&self) {
+        self.registers.cr2.modify(CR2::TXDMAEN::CLEAR);
+    }
+
+    fn enable_rx(&self) {
+        self.registers.cr2.modify(CR2::RXDMAEN::SET);
+    }
+
+    fn disable_rx(&self) {
+        self.registers.cr2.modify(CR2::RXDMAEN::CLEAR);
+    }
+
+    fn read_write_bytes(
+        &self,
+        write_buffer: Option<&'static mut [u8]>,
+        read_buffer: Option<&'static mut [u8]>,
+        len: usize,
+    ) -> ReturnCode {
+        if write_buffer.is_none() && read_buffer.is_none() {
+            return ReturnCode::EINVAL;
+        }
+
+        self.hold_low();
+
+        let mut count: usize = len;
+        write_buffer
+            .as_ref()
+            .map(|buf| count = cmp::min(count, buf.len()));
+        read_buffer
+            .as_ref()
+            .map(|buf| count = cmp::min(count, buf.len()));
+
+        self.dma_len.set(count);
```

Comment:
hrm, it is not necessarily true that `Some(write_buffer)` and `Some(read_buffer)` is of the same length here.

Can't elements be missed if for example `write_buffer > rx_buffer`?

Reply:
But you _have_ to do that, don't you? Since SPI is symmetrical, you have to have both a read and write buffer of the same length, and the only safe option is to take the minimum of the two.

Reply:
I think @alevy is correct.

Reply:
Cool, I just wanted the raise the question and confirm that it was intended so makes sense to me.

> and the only safe option is to take the minimum of the two.

Well, that is true unless one of them is `None`, then we just set it to `min(Some(buf.len()), len)` which makes sense.  If we would map None -> 0 then nothing would be written/read